pkg/k8s/client.go

```diff
@@ -0,0 +1,67 @@
+package k8s
+
+import (
+	"context"
+	"fmt"
+
+	apiv1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/util/retry"
+)
+
+type Client struct {
+	Clientset kubernetes.Interface
+}
+
+func (c Client) ApplySecret(ctx context.Context, secret *apiv1.Secret) error {
+	_, err := c.createSecret(ctx, secret)
+	if err != nil {
+		err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			// Retrieve the latest version of Secret before attempting update
+			// RetryOnConflict uses exponential backoff to avoid exhausting the apiserver
+			gs, err := c.getSecret(ctx, secret)
+			if err != nil {
+				return err
+			}
+
+			if err := c.updateSecret(ctx, gs); err != nil {
+				return err
+			}
+
+			return nil
+		})
+
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (c Client) createSecret(ctx context.Context, secret *apiv1.Secret) (*apiv1.Secret, error) {
+	if _, err := c.Clientset.CoreV1().Secrets(secret.Namespace).Create(ctx, secret, metav1.CreateOptions{}); err != nil {
+		return nil, fmt.Errorf("secret probably already exists %s", err)
```

Comment:
At this point no. Worth the conversation to determine if it would be useful here.  I have not investigated vaults api to see if there are any useful error types that it returns.
 
A nice write up on when / why to use your own error types, I generally agree with dc on this but its a loosely held opinion.  https://dave.cheney.net/2016/04/27/dont-just-check-errors-handle-them-gracefully